File: app/src/main/kotlin/com/openclaw/ghostcrab/data/impl/AIRecommendationServiceImpl.kt
package com.openclaw.ghostcrab.data.impl

import com.openclaw.ghostcrab.data.api.dto.AIContextDto
import com.openclaw.ghostcrab.data.api.dto.AIRecommendationRequestDto
import com.openclaw.ghostcrab.domain.exception.AIQuotaExceededException
import com.openclaw.ghostcrab.domain.exception.AIServiceUnavailableException
import com.openclaw.ghostcrab.domain.exception.GatewayApiException
import com.openclaw.ghostcrab.domain.model.AIRecommendation
import com.openclaw.ghostcrab.domain.model.GatewayConnection
import com.openclaw.ghostcrab.domain.model.RecommendationContext
import com.openclaw.ghostcrab.domain.model.SuggestedChange
import com.openclaw.ghostcrab.domain.repository.AIRecommendationService
import kotlinx.serialization.json.JsonObject

/**
 * Ktor-backed [AIRecommendationService] that proxies queries to the gateway's AI skill.
 *
 * [isAvailable] is a capability check against the cached connection state — no network call.
 * [getRecommendation] POSTs to `POST /api/ai/recommend` on the connected gateway.
 *
 * @param connectionManager Provides the active [OpenClawApiClient] and connection state.
 */
class AIRecommendationServiceImpl(
    private val connectionManager: GatewayConnectionManagerImpl,
) : AIRecommendationService {

    /**
     * Returns `true` if the currently-connected gateway advertises the `skill-ai-recommend`
     * capability in its `/status` response.
     *
     * Reads the in-memory connection state — no network call is made.
     *
     * @return `true` when [GatewayConnection.Connected.capabilities] contains `"skill-ai-recommend"`.
     */
    override suspend fun isAvailable(): Boolean {
        val state = connectionManager.connectionState.value
        return state is GatewayConnection.Connected &&
            SKILL_KEY in state.capabilities
    }

    /**
     * Submits [query] and auto-collected [context] to `POST /api/ai/recommend`.
     *
     * @param query Free-text user query.
     * @param context Session context assembled by the ViewModel (config + hardware + active model).
     * @return [AIRecommendation] mapped from the gateway DTO.
     * @throws AIServiceUnavailableException if the gateway returns 404 (skill not installed).
     * @throws AIQuotaExceededException if the gateway returns 429 (rate limit).
     * @throws com.openclaw.ghostcrab.domain.exception.GatewayException on other network errors.
     * @throws IllegalStateException if no active connection exists.
     */
    override suspend fun getRecommendation(
        query: String,
        context: RecommendationContext,
    ): AIRecommendation {
        val client = connectionManager.requireClient()
        val request = AIRecommendationRequestDto(
            query = query,
            context = AIContextDto(
                activeConfig = JsonObject(context.activeConfig.sections),
                hardwareInfo = context.hardwareInfo,
                activeModelId = context.activeModelId,
            ),
        )
        return try {
            client.getAIRecommendation(request).let { dto ->
                AIRecommendation(
                    query = query,
                    recommendation = dto.recommendation,
                    suggestedChanges = dto.suggestedChanges.map { change ->
                        SuggestedChange(
                            section = change.section,
                            key = change.key,
                            currentValue = change.currentValue,
                            suggestedValue = change.suggestedValue,
                            rationale = change.rationale,
                        )
                    },
                )
            }
        } catch (e: GatewayApiException) {
            when (e.statusCode) {
                404 -> throw AIServiceUnavailableException(client.baseUrl)
                429 -> throw AIQuotaExceededException(client.baseUrl)
                else -> throw e
            }
        }
    }

}

private const val SKILL_KEY = "skill-ai-recommend"

File: ios/GhostCrab/Data/Impl/AIRecommendationServiceImpl.swift
import Foundation

/// URLSession-backed ``AIRecommendationService`` that proxies queries to the
/// gateway's AI skill.
///
/// ``isAvailable()`` is a capability check against the cached connection state
/// — no network call. ``getRecommendation(query:context:)`` POSTs to
/// `POST /api/ai/recommend` on the connected gateway.
///
/// Direct port of `AIRecommendationServiceImpl.kt`.
///
/// **Note on error mapping**: the Kotlin source catches `GatewayApiException`
/// with status 404/429 and re-throws domain-specific subclasses. The Swift
/// ``OpenClawAPIClient`` already throws ``GatewayError/aiServiceUnavailable``
/// and ``GatewayError/aiQuotaExceeded`` directly for those status codes — so
/// here we just let them propagate. The visible-from-callers behaviour is
/// identical.
public final class AIRecommendationServiceImpl: AIRecommendationService, Sendable {

    private static let skillKey = "ai.recommend"

    private let connectionManager: GatewayConnectionManagerImpl

    /// - Parameter connectionManager: Provides the active ``OpenClawAPIClient``
    ///   and connection state.
    public init(connectionManager: GatewayConnectionManagerImpl) {
        self.connectionManager = connectionManager
    }

    /// Returns `true` if the currently-connected gateway advertises the
    /// `ai.recommend` capability in its `/status` response.
    ///
    /// Reads the in-memory connection state — no network call is made.
    public func isAvailable() async -> Bool {
        let state = await connectionManager.currentState
        if case let .connected(_, _, _, _, _, capabilities, _, _) = state {
            return capabilities.contains(Self.skillKey)
        }
        return false
    }

    /// Submits `query` and auto-collected `context` to `POST /api/ai/recommend`.
    ///
    /// - Parameters:
    ///   - query: Free-text user query.
    ///   - context: Session context assembled by the ViewModel.
    /// - Returns: ``AIRecommendation`` mapped from the gateway DTO.
    /// - Throws: ``GatewayError/aiServiceUnavailable(url:)`` on 404
    ///   (skill not installed).
    /// - Throws: ``GatewayError/aiQuotaExceeded(url:)`` on 429 (rate limit).
    /// - Throws: ``GatewayError`` on other network errors.
    /// - Throws: ``GatewayError/unreachable(url:underlying:)`` if no active
    ///   connection exists.
    public func getRecommendation(
        query: String,
        context: RecommendationContext
    ) async throws -> AIRecommendation {
        let client = try await connectionManager.requireClient()
        let request = AIRecommendationRequestDTO(
            query: query,
            context: AIContextDTO(
                activeConfig: context.activeConfig.sections,
                hardwareInfo: context.hardwareInfo,
                activeModelId: context.activeModelId
            )
        )
        let dto = try await client.getAIRecommendation(request)
        return AIRecommendation(
            query: query,
            recommendation: dto.recommendation,
            suggestedChanges: dto.suggestedChanges.map { change in
                SuggestedChange(
                    section: change.section,
                    key: change.key,
                    currentValue: change.currentValue,
                    suggestedValue: change.suggestedValue,
                    rationale: change.rationale
                )
            }
        )
    }
}
